### PR TITLE
lint: add bad opt access internal lint

### PR DESCRIPTION
**Pull request by davidtwco**
Prompted by [Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/sess.2Ecrate_types.28.29.20vs.20sess.2Eopts.2Ecrate_types/near/290682847).

Some command-line options accessible through `sess.opts` are best accessed through wrapper functions on `Session`, `TyCtxt` or otherwise, rather than through field access on the option struct in the `Session`.

Adds a new lint which triggers on those options that should be accessed through a wrapper function so that this is prohibited. Options are annotated with a new attribute `rustc_lint_opt_deny_field_access` which can specify the error message (i.e. "use this other function instead") to be emitted.

A simpler alternative would be to simply rename the options in the option type so that it is clear they should not be used, however this doesn't prevent uses, just discourages them. Another alternative would be to make the option fields private, and adding accessor functions on the option types, however the wrapper functions sometimes rely on additional state from `Session` or `TyCtxt` which wouldn't be available in an function on the option type, so the accessor would simply make the field available and its use would be discouraged too.

**Leave a comment if there's an option I should add this to.**

### CODE REVIEW

**Comment by bjorn3** [comment]
Maybe also enable this lint for cli_forced_codegen_units, verify_llvm_ir, lto, thinlto and plt? Or even for all fields which have a wrapper, no matter how small?

**Comment by RalfJung** [comment]
> A simpler alternative would be to simply rename the options 

An even simpler alternative would be to make these fields inaccessible using standard Rust visibility restrictions?

**Comment by bjorn3** [comment]
Sometimes they need to be accessed without an available `Session`. For example https://github.com/rust-lang/rust/pull/99710/files#diff-96da842aaf6582a52b9f23b4b353ef42214149dfb874819da8b616bacfb4b43eR564 for crate_types. Or every option for custom drivers that want to change options. (for example when cg_clif used a custom driver, it set panic=abort from inside the driver)

**Comment by davidtwco** [comment]
> > A simpler alternative would be to simply rename the options
> 
> An even simpler alternative would be to make these fields inaccessible using standard Rust visibility restrictions?

That's the second alternative I listed, it would absolutely work, and we could add our wrapper functions to `Options`/`CodegenOptions`/`DebuggingOptions`. But as soon as those wrapper functions need to make use of something from `Session` or `TyCtxt`, then we either need to pass those in (possible for `Session`, but probably not for `TyCtxt`), or the wrapper function just becomes an accessor that another wrapper on `Session`/`TyCtxt` uses - in which case it could be used accidentally just like a public field. A lint is a bit more machinery but it keeps the option accessible for the wrapper function, wherever that exists, but also prevents it from being used otherwise. I might be missing something though.

File: compiler/rustc_interface/src/interface.rs [hunk 1]
```diff
@@ -329,6 +329,8 @@ pub fn create_compiler_and_run<R>(config: Config, f: impl FnOnce(&Compiler) -> R
     })
 }
 
+// JUSTIFICATION: interface fn
```
[the remark this answers is not in the document]
Comment (davidtwco) [reply compiler/rustc_interface/src/interface.rs:332]: Apologies, I'll improve this comment. Just that I would expect the functions in `rustc_interface` to access some options directly because they might be running before the `Session` is constructed, or something along those lines.

File: compiler/rustc_session/src/session.rs [hunk 2]
```diff
@@ -583,35 +583,73 @@ impl Session {
...
     pub fn time_passes(&self) -> bool {
-        self.opts.unstable_opts.time_passes || self.opts.unstable_opts.time
+        self.opts.time_passes()
     }
+
+    /// Returns `true` if internal lints should be added to the lint store - i.e. if
+    /// `-Zunstable-options` is provided and this isn't rustdoc (internal lints can trigger errors
+    /// to be emitted under rustdoc).
+    pub fn enable_internal_lints(&self) -> bool {
+        self.unstable_options() && !self.opts.actually_rustdoc
```
Comment (lcnr) [inline compiler/rustc_session/src/session.rs:601]: with this, can you remove the `FIXME` in `fn typeck_results_of_method_fn`, i think that should not be necessary anymore

File: compiler/rustc_session/src/session.rs [hunk 3]
```diff
@@ -583,35 +583,73 @@ impl Session {
...
+    }
+
+    // JUSTIFICATION: defn of the suggested wrapper fn
+    #[cfg_attr(not(bootstrap), allow(rustc::bad_opt_access))]
```
Comment (lcnr) [inline compiler/rustc_session/src/session.rs:605]: can you instead move all of the suggested wrapper functions into a separate impl to not repeat this?

File: compiler/rustc_session/src/session.rs [hunk 4]
```diff
@@ -738,6 +792,8 @@ impl Session {
         let found_negative = requested_features.clone().any(|r| r == "-crt-static");
         let found_positive = requested_features.clone().any(|r| r == "+crt-static");
 
+        // JUSTIFICATION: necessary use of crate_types directly
```
Comment (lcnr) [inline compiler/rustc_session/src/session.rs:795]: why is that necessary?

File: compiler/rustc_session/src/session.rs [hunk 5]
```diff
@@ -1092,20 +1168,31 @@ impl Session {
...
+    #[cfg_attr(not(bootstrap), allow(rustc::bad_opt_access))]
     pub fn link_dead_code(&self) -> bool {
         self.opts.cg.link_dead_code.unwrap_or(false)
     }
 
+    // JUSTIFICATION: defn of the suggested wrapper fn
+    #[cfg_attr(not(bootstrap), allow(rustc::bad_opt_access))]
     pub fn instrument_coverage(&self) -> bool {
-        self.opts.instrument_coverage()
+        self.opts.cg.instrument_coverage.unwrap_or(InstrumentCoverage::Off)
```
Comment (lcnr) [inline compiler/rustc_session/src/session.rs:1180]: maybe put the `self.opts.cg.instrument_coverage.unwrap_or(InstrumentCoverage::Off)` as a method of `CodegenOptions` or as a new private function? :thinking:

**Review by lcnr — COMMENTED** [review]
a few nits, then r=me

haven't really checked the fields which have this lint, but even if one of the fields is incorrectly linted, that shouldn't hurt too much

File: compiler/rustc_session/src/session.rs (hunk 4 above)
Comment (bjorn3) [reply compiler/rustc_session/src/session.rs:795]: It is called from `validate_commandline_args_with_session_available` at which point `sess.crate_types` is not yet initialized. There is a FIXME a couple lines below about this.

File: compiler/rustc_session/src/session.rs (hunk 2 above)
Comment (davidtwco) [reply compiler/rustc_session/src/session.rs:595]: Fixed

File: compiler/rustc_session/src/session.rs (hunk 3 above)
Comment (davidtwco) [reply compiler/rustc_session/src/session.rs]: Fixed

File: compiler/rustc_session/src/session.rs (hunk 4 above)
Comment (davidtwco) [reply compiler/rustc_session/src/session.rs]: I've modified the comment to reference the FIXME below.

File: compiler/rustc_session/src/session.rs (hunk 5 above)
Comment (davidtwco) [reply compiler/rustc_session/src/session.rs]: Fixed

**Comment by jyn514** [comment]
> Another alternative would be to make the option fields private, and adding accessor functions on the option types, however the wrapper functions sometimes rely on additional state from `Session` or `TyCtxt` which wouldn't be available in an function on the option type, so the accessor would simply make the field available and its use would be discouraged too.

Can we instead move those fields to be defined in the same crate as Session? This seems like an improvement over the status quo, but at the cost of increased complexity ...

**Comment by davidtwco** [comment]
> > Another alternative would be to make the option fields private, and adding accessor functions on the option types, however the wrapper functions sometimes rely on additional state from `Session` or `TyCtxt` which wouldn't be available in an function on the option type, so the accessor would simply make the field available and its use would be discouraged too.
> 
> Can we instead move those fields to be defined in the same crate as Session? This seems like an improvement over the status quo, but at the cost of increased complexity ...

They already are, we could use `pub(crate)` (honestly, that skipped my mind entirely when writing this, for some reason, I only considered `pub`). If we want to do that instead then I don't mind `r=me`'ing a PR that removes this and does that instead.

**Comment by bjorn3** [comment]
`pub(crate)` wouldn't allow options to be changed in `rustc_driver::Callbacks::config`.

**Comment by Mark-Simulacrum** [comment]
It also doesn't prevent other accessors not using the accessor for this field - I think to do that we'd need wrapper structs in private modules or something, which feels ungreat. I think the lint is a pretty OK method of achieving our goals here.
